sql/inspect: introduce interface for INSPECT check infrastructure

spilchen opened the pull request:
Add a new interface to support INSPECT operations that perform checks emitting rows for each issue found. This interface follows a row iterator pattern, enabling future implementation of specific checks.

Informs: #146863
Epic: CRDB-30356
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/149423)
<!-- Reviewable:end -->

rafiss reviewed (changes requested):
I just had minor nits

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/149423#-:-OUerZihCVqYRPGCgVmT:b-ivl972)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`pkg/sql/inspect/log_sink.go` line 21 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/149423#-OUer6gLAPIu5Y0PpGyk:-OUer6gMCvPUDlJE26rs:b-hafepu) ([raw file](https://github.com/cockroachdb/cockroach/blob/81483a82be6562dca5937b4814fc40bd441748c7/pkg/sql/inspect/log_sink.go#L21)):*
> ```Go
> // logIssue implements the inspectLogger interface.
> func (c *logSink) logIssue(ctx context.Context, issue *inspectIssue) error {
> 	log.Errorf(ctx, "inspect issue: %+v", issue)
> ```

nit: consider having `*inspectIssue` implement `redact.SafeFormatter` interface so these logs are redacted correctly.
___
*[`pkg/sql/inspect/issue.go` line 37 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/149423#-OUeqz9g7XChv_Xrfb4t:-OUeqz9g7XChv_Xrfb4u:bf478i2) ([raw file](https://github.com/cockroachdb/cockroach/blob/81483a82be6562dca5937b4814fc40bd441748c7/pkg/sql/inspect/issue.go#L37)):*
> ```Go
> 	// Details contains additional structured metadata describing the issue.
> 	// The contents vary by check type.
> 	Details map[string]interface{}
> ```

nit: do we want this key to be a `RedactableString`?


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/149423)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained

___
*[`pkg/sql/inspect/log_sink.go` line 21 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/149423#-OUer6gLAPIu5Y0PpGyk:-OUfNTTi-H_M0l6mlxPa:b-ovwaw7) ([raw file](https://github.com/cockroachdb/cockroach/blob/81483a82be6562dca5937b4814fc40bd441748c7/pkg/sql/inspect/log_sink.go#L21)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: consider having `*inspectIssue` implement `redact.SafeFormatter` interface so these logs are redacted correctly.
</blockquote></details>

Good idea.


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
lgtm!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/149423#-:-OUkZ3NH8ZxyICcDWgTO:b-7nbn20)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`pkg/sql/inspect/issue_test.go` line 46 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/149423#-OUkYuol3oQROJU3Vl6S:-OUkYuol3oQROJU3Vl6T:b-203atb) ([raw file](https://github.com/cockroachdb/cockroach/blob/29f49275ae46070e1f3d2465763f4b0bf7f3b3f1/pkg/sql/inspect/issue_test.go#L46)):*
> ```Go
> 				Details: map[redact.RedactableString]interface{}{
> 					"field1": 10,
> 					"field2": "strval",
> ```

super nit: if a `RedactableString` (or any type that implements `SafeFormatter`) is in the value of the map, does the redaction apply as we expect?


<!-- Sent from Reviewable.io -->

spilchen committed:
sql/inspect: introduce interface for INSPECT check infrastructure

Add a new interface to support INSPECT operations that perform checks emitting
rows for each issue found. This interface follows a row iterator pattern,
enabling future implementation of specific checks.

Informs: #146863
Epic: CRDB-30356
Release note: none

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/149423)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/sql/inspect/issue_test.go` line 46 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/149423#-OUkYuol3oQROJU3Vl6S:-OUlenqa88Ys7X56NVzj:b-2rwrhv) ([raw file](https://github.com/cockroachdb/cockroach/blob/29f49275ae46070e1f3d2465763f4b0bf7f3b3f1/pkg/sql/inspect/issue_test.go#L46)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

super nit: if a `RedactableString` (or any type that implements `SafeFormatter`) is in the value of the map, does the redaction apply as we expect?
</blockquote></details>

Good suggestion. I added a test to confirm. If the value is a `RedactableString`, then it won't get redacted out, which I think is the behaviour that we want.


<!-- Sent from Reviewable.io -->
